mne/stats/regression.py

```diff
@@ -134,3 +143,208 @@ def _fit_lm(data, design_matrix, names):
+    events[:, 0] /= decim
+
+    conds = list(event_id)
+    if covariates is not None:
+        conds += list(covariates)
+
+    # time windows (per event type)
+    if isinstance(tmin, (float, int)):
+        tmin = dict((cond, int(tmin * sfreq)) for cond in conds)
+    else:
+        tmin = dict((cond, int(tmin.get(cond, -.1) * sfreq))
+                    for cond in conds)
+    if isinstance(tmax, (float, int)):
+        tmax = dict((cond, int((tmax * sfreq) + 1.)) for cond in conds)
+    else:
+        tmax = dict((cond, int((tmax.get(cond, 1.) * sfreq) + 1))
+                    for cond in conds)
+
+    # Construct predictor matrix
+    # We do this by creating one array per event type, shape lags * samples
+    # (where lags depends on tmin/tmax and can be different for different
+    # event types). Columns correspond to predictors, predictors correspond to
+    # time lags. Thus, each array is mostly sparse, with one diagonal of 1s
+    # per event (for binary predictors).
+
+    # This should probably be improved (including making it more robust to
+    # high frequency data) by operating on sparse matrices. As-is, high
+    # sampling rates plus long windows blow up the system due to the inversion
+    # of massive matrices.
+    # Furthermore, assigning to a preallocated array would be faster.
+
+    cond_length = dict()
+    pred_arrays = list()
+    for cond in conds:
+
+        # create the first row and column to be later used by toeplitz to build
+        # the full predictor matrix
+        tmin_, tmax_ = tmin[cond], tmax[cond]
+        n_lags = int(tmax_ - tmin_)
+        samples, lags = np.zeros(len(times)), np.zeros(n_lags)
+
+        if cond in event_id:  # for binary predictors
+            ids = ([event_id[cond]] if isinstance(event_id[cond], int)
+                   else event_id[cond])
+            samples[events[np.in1d(events[:, 2], ids), 0] + int(tmin_)] = 1
+
+        else:  # for predictors from covariates, e.g. continuous ones
+            if len(covariates[cond]) != len(events):
+                error = """Condition {} from ```covariates``` is
+                        not the same length as ```events```""".format(cond)
+                raise ValueError(error)
+            for tx, v in zip(events[:, 0], covariates[cond]):
+                samples[tx + int(tmin_)] = np.float(v)
+
+        cond_length[cond] = len(np.nonzero(samples))
+
+        # this is the magical part (thanks to Marijn van Vliet):
+        # use toeplitz to construct series of diagonals
+        pred_arrays.append(linalg.toeplitz(samples, lags))
+
+    big_arr = np.hstack(pred_arrays).T
+    # find only those positions where at least one predictor isn't 0
+    has_val = np.asarray([len(np.nonzero(line)[0]) > 0
+                          for line in big_arr.T])
+
+    # additionally, reject positions based on extreme steps in the data
+    if reject is not False:
+        from ..utils import _reject_data_segments
```

Comment:
In general it's cleaner to put everything at the top. If I want to answer the question "what functions and classes will be used by the functions in this module?" it's easier to answer. There are a few reasons to nest:

1. Speed up imports
2. Optional/soft requirements
3. Circular imports

None of those really apply here. `mne.utils` will be imported as soon as `mne` gets imported so it won't affect import speed, and the other two aren't relevant here. Make sense?